pandas/tools/merge.py

```diff
@@ -47,9 +54,56 @@ class MergeError(ValueError):
     pass
 
 
-def ordered_merge(left, right, on=None, left_by=None, right_by=None,
+def _groupby_and_merge(by, left, right, _merge_pieces):
+    """
+    groupby & merge
+    we are always performing a left-by type operation
+    """
+
+    pieces = []
+    if not isinstance(by, (list, tuple)):
+        by = [by]
+
+    lby = left.groupby(by, sort=False)
+
+    # if we can groupby the rhs
+    # then we can get vastly better perf
+    try:
+        rby = right.groupby(by, sort=False)
+    except KeyError:
+        rby = None
+    for key, lhs in lby:
+
+        if rby is None:
+            rhs = right
+        else:
+            try:
+                rhs = right.iloc[rby.indices[key]]
+            except KeyError:
+                # key doesn't exist in left
```

Comment:
Why wouldn't the key exist in the left if the failure is from a lookup on the right?